File: protocol/src/main/java/spp/protocol/probe/command/LiveInstrumentContext.java
package spp.protocol.probe.command;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import io.vertx.core.json.Json;

import java.io.Serializable;
import java.util.*;
import java.util.stream.Collectors;

@JsonIgnoreProperties(ignoreUnknown = true)
public class LiveInstrumentContext implements Serializable {

    private Set<String> instruments = new HashSet<>();
    private Set<String> locations = new HashSet<>();

    public LiveInstrumentContext() {
    }

    public List<String> getLiveInstruments() {
        return new ArrayList<>(instruments);
    }

    public LiveInstrumentContext addLiveInstrument(Object liveInstrument) {
        instruments.add(Json.encode(liveInstrument));
        return this;
    }

    public LiveInstrumentContext addLiveInstrument(String liveInstrument) {
        instruments.add(liveInstrument);
        return this;
    }

    public LiveInstrumentContext addLiveInstruments(Collection<String> liveInstruments) {
        this.instruments.addAll(liveInstruments);
        return this;
    }

    public <T> List<T> getLiveInstrumentsCast(Class<T> clazz) {
        return instruments.stream().map(it -> Json.decodeValue(it, clazz)).collect(Collectors.toList());
    }

    public List<String> getLocations() {
        return new ArrayList<>(locations);
    }

    public void addLocation(Object location) {
        locations.add(Json.encode(location));
    }

    @SuppressWarnings("unused")
    public void addLocations(Collection<String> locations) {
        this.locations = new HashSet<>(locations);
    }

    @SuppressWarnings("unused")
    public <T> List<T> getLocationsCast(Class<T> clazz) {
        return locations.stream().map(it -> Json.decodeValue(it, clazz)).collect(Collectors.toList());
    }

    @SuppressWarnings("unused")
    public void setLiveInstruments(Set<String> instruments) {
        this.instruments = instruments;
    }

    @SuppressWarnings("unused")
    public void setLocations(Set<String> locations) {
        this.locations = locations;
    }

    @Override
    public String toString() {
        return "LiveInstrumentContext{" +
                "instruments=" + instruments +
                ", locations=" + locations +
                '}';
    }
}

File: protocol/src/main/kotlin/spp/protocol/probe/command/LiveInstrumentContext.kt
package spp.protocol.probe.command

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import io.vertx.core.json.Json
import java.io.Serializable
import java.util.stream.Collectors

//todo: treat this as a regular data class
@JsonIgnoreProperties(ignoreUnknown = true)
data class LiveInstrumentContext(
    var instruments: MutableSet<String> = HashSet(),
    var locations: MutableSet<String> = HashSet()
) : Serializable {

    @get:JsonIgnore
    val liveInstruments: List<String>
        get() = instruments.toList()

    fun addLiveInstrument(liveInstrument: Any): LiveInstrumentContext {
        instruments.add(Json.encode(liveInstrument))
        return this
    }

    fun addLiveInstrument(liveInstrument: String): LiveInstrumentContext {
        instruments.add(liveInstrument)
        return this
    }

    fun addLiveInstruments(liveInstruments: Collection<String>): LiveInstrumentContext {
        instruments.addAll(liveInstruments)
        return this
    }

    fun <T> getLiveInstrumentsCast(clazz: Class<T>): List<T> {
        return instruments.stream().map { Json.decodeValue(it, clazz) }.collect(Collectors.toList())
    }

    fun addLocation(location: Any) {
        locations.add(Json.encode(location))
    }

    fun <T> getLocationsCast(clazz: Class<T>): List<T> {
        return locations.stream().map { Json.decodeValue(it, clazz) }.collect(Collectors.toList())
    }
}
